driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -167,10 +174,47 @@ private OidcCallback getOidcCallbackMechanismProperty(final String key) {
                 .getMechanismProperty(key, null);
     }
 
-    @Nullable
     private OidcCallback getRequestCallback() {
-        OidcCallback machine = getOidcCallbackMechanismProperty(OIDC_CALLBACK_KEY);
-        return machine != null ? machine : getOidcCallbackMechanismProperty(OIDC_HUMAN_CALLBACK_KEY);
+        String environment = getEnvironmentName(getMongoCredential());
+        OidcCallback machine;
+        if (TEST_ENVIRONMENT.equals(environment)) {
+            machine = getTestCallback();
+        } else if (AZURE_ENVIRONMENT.equals(environment)) {
+            machine = getAzureCallback(getMongoCredential());
+        } else if (GCP_ENVIRONMENT.equals(environment)) {
+            machine = getGcpCallback(getMongoCredential());
+        } else {
+            machine = getOidcCallbackMechanismProperty(OIDC_CALLBACK_KEY);
+        }
+        OidcCallback human = getOidcCallbackMechanismProperty(OIDC_HUMAN_CALLBACK_KEY);
+        return machine != null ? machine : assertNotNull(human);
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    public static OidcCallback getTestCallback() {
+        return (context) -> {
+            String accessToken = readTokenFromFile();
+            return new OidcCallbackResult(accessToken);
+        };
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    public static OidcCallback getAzureCallback(final MongoCredential credential) {
+        return (context) -> {
+            String resource = assertNotNull(credential.getMechanismProperty(TOKEN_RESOURCE_KEY, null));
+            String objectId = credential.getUserName();
```

Comment:
This is actually the clientId

Reply:
Hmm, my ensuing fetchAzure... call is calling:

```
    public static CredentialInfo fetchAzureCredentialInfo(final String resource, @Nullable final String objectId) {
        String endpoint = "http://169.254.169.254:80"
                + "/metadata/identity/oauth2/token?api-version=2018-02-01"
                + "&resource=" + resource
                + (objectId == null ? "" : "&object_id=" + objectId);
```
This must be incorrect, since the spec says to use:
```
http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=<resource>&client_id=<client_id>
```
I'll update this, but why do the tests still pass, and was this changed from object_id?

Reply:
It was changed from object_id to align with the Azure SDK that uses client_id.